File: orchestrator/internal/handoff/handoff.go
package handoff

import (
	"os"
	"path/filepath"
	"strings"
)

type Note struct {
	Source  string
	Content string
}

func Collect(root string) ([]Note, error) {
	var notes []Note
	info, err := os.Stat(root)
	if err != nil || !info.IsDir() {
		return notes, nil
	}
	err = filepath.Walk(root, func(path string, fi os.FileInfo, err error) error {
		if err != nil {
			return nil
		}
		if fi.IsDir() {
			return nil
		}
		if strings.ToUpper(filepath.Base(path)) == "HANDOFF.MD" {
			data, err := os.ReadFile(path)
			if err != nil {
				return nil
			}
			notes = append(notes, Note{
				Source:  path,
				Content: string(data),
			})
		}
		return nil
	})
	return notes, err
}

func FormatNotes(notes []Note) string {
	if len(notes) == 0 {
		return ""
	}
	var parts []string
	for _, n := range notes {
		parts = append(parts, "## Handoff: "+n.Source+"\n\n"+n.Content)
	}
	return strings.Join(parts, "\n\n---\n\n")
}

File: orchestrator/handoff.py
from dataclasses import dataclass
from pathlib import Path

from orchestrator import log

logger = log.get("handoff")


@dataclass
class HandoffNote:
    source: str
    content: str


def collect(root: str) -> list[HandoffNote]:
    notes: list[HandoffNote] = []
    base = Path(root)
    if not base.exists():
        return notes
    for path in sorted(base.rglob("HANDOFF.md")):
        if path.is_file():
            notes.append(HandoffNote(
                source=str(path),
                content=path.read_text(encoding="utf-8"),
            ))
    logger.info("collected %d handoff notes from %s", len(notes), root)
    return notes


def format_notes(notes: list[HandoffNote]) -> str:
    if not notes:
        return ""
    parts: list[str] = []
    for n in notes:
        parts.append(f"## Handoff: {n.source}\n\n{n.content}")
    return "\n\n---\n\n".join(parts)
